File: Problem_030/Problem_030.py
def digit_nth_powers(n):
    total_sum = 0
    # The result of 5, 6, and 7 nines was the same
    # Therefore only iterating up to 9^5 * 5
    for num in range(2, 9**n * n):
        sum = 0
        for digit in str(num):
            sum += int(digit)**n
        if sum == num:
            total_sum += num

    return total_sum


print(digit_nth_powers(5))

File: Problem_030/Problem_030.cpp
#include <iostream>
using namespace std;

int power(int n, int p)
{
  int result = n;
  for (int i = 1; i < p; i++)
  {
    result *= n;
  }
  return result;
}

int digit_nth_powers(int n)
{
  int total_sum = 0;

  for (int num_const = 2; num_const <= power(9, n) * n; num_const++)
  {
    int num = num_const;
    int sum = 0;
    while (num != 0)
    {
      int digit = num % 10;
      int digit_to_nth_power = power(digit, n);
      sum += digit_to_nth_power;
      num /= 10;
    }
    if (sum == num_const)
    {
      total_sum += num_const;
    }
  }

  return total_sum;
}

int main()
{
  cout << digit_nth_powers(5) << endl;
  return 0;
}
